File: templating/core/src/main/python/jsonteng/tags/tag_base.py
# Copyright 2019 VMware, Inc.
# SPDX-License-Indentifier: Apache-2.0

import abc

from ..exception import TemplateEngineException


class TagBase(object):
    """
    Base class of tag classes.
    """
    __metaclass__ = abc.ABCMeta
    TAG_NONE = type('TagNone', (), {})()

    def __init__(self, tag_resolver):
        """
        Base tag constructor.
        :param tag_resolver: Rag resolver to be used by the tag.
        """
        self._tag_resolver = tag_resolver

    @abc.abstractmethod
    def process(self, tag_tokens, binding_data_list):
        """
        Process a tag.
        :param tag_tokens: Tag arguments.
        :type tag_tokens: 'list'
        :param binding_data_list: Binding data used during the processing.
        :type binding_data_list: 'list'
        :return: JSON object
        :rtype: JSON object
        """
        raise NotImplementedError

    @staticmethod
    def safe_eval(expr):
        """
        A safe eval that is limited to simple expressions.
        :param expr: A Python boolean expression
        :type expr: 'str'
        :return: result of expression evaluation
        :rtype: 'bool'
        """
        result = eval(expr, {"__builtins__": None}, {})
        if not isinstance(result, bool):
            raise TemplateEngineException(
                "Expression {} is not a boolean type".format(expr))
        return result

File: templating/core/src/main/go/src/jsonteng/tags/tag_base.go
// Copyright 2019 VMware, Inc.
// SPDX-License-Indentifier: Apache-2.0

package tags

import (
	"fmt"
	"io"
	"jsonteng/errors"
	intfs "jsonteng/interfaces"
	extTags "jsonteng_contribs/tags"
	"os/exec"
	"strings"
)

// BuildTagMap builds a map containing intrinsic tags.
func BuildTagMap(tagResolver intfs.TagResolver) *map[string]intfs.TagProcessor {
	tagMap := make(map[string]intfs.TagProcessor)
	atTag := atTag{}
	atTag.init(tagResolver)
	tagMap[atTag.Name] = &atTag
	existsTag := existsTag{}
	existsTag.init(tagResolver)
	tagMap[existsTag.Name] = &existsTag
	foreachTag := foreachTag{}
	foreachTag.init(tagResolver)
	tagMap[foreachTag.Name] = &foreachTag
	lenTag := lenTag{}
	lenTag.init(tagResolver)
	tagMap[lenTag.Name] = &lenTag
	oneofTag := oneofTag{}
	oneofTag.init(tagResolver)
	tagMap[oneofTag.Name] = &oneofTag
	toboolTag := toboolTag{}
	toboolTag.init(tagResolver)
	tagMap[toboolTag.Name] = &toboolTag
	tofloatTag := tofloatTag{}
	tofloatTag.init(tagResolver)
	tagMap[tofloatTag.Name] = &tofloatTag
	tointTag := tointTag{}
	tointTag.init(tagResolver)
	tagMap[tointTag.Name] = &tointTag
	tonullTag := tonullTag{}
	tonullTag.init(tagResolver)
	tagMap[tonullTag.Name] = &tonullTag

	extMap := extTags.BuildTagMap(tagResolver)
	for k, v := range extMap {
		tagMap[k] = v
	}
	return &tagMap
}

func safeEval(exprObj interface{}) (bool, error) {
	expr, ok := exprObj.(string)
	if !ok {
		if expr, ok := exprObj.(bool); ok {
			return expr, nil
		}
		errl := errors.GenericError
		errl.Message = fmt.Sprintf("SafeEval: Invalid expr type. %v", exprObj)
		return false, &errl
	}
	if expr == "True" {
		return false, nil
	} else if expr == "False" {
		return false, nil
	}
	script := fmt.Sprintf("print(eval('%s', {'__builtins__': None}, {}))", expr)
	cmd := exec.Command("python", "-")
	stdin, err := cmd.StdinPipe()
	if err != nil {
		errl := errors.GenericError
		errl.Message = fmt.Sprintf("SafeEval: Expression %v evaluation failed due to stdin pipe closed. %v", exprObj)
		return false, &errl
	}
	go func() {
		defer stdin.Close()
		io.WriteString(stdin, script)
	}()
	result, err := cmd.Output()
	if err != nil {
		errl := errors.GenericError
		errl.Message = fmt.Sprintf("SafeEval: Expression %v evaluation failed. %v", exprObj, err)
		return false, &errl
	}
	resultStr := strings.TrimSpace(string(result))
	if strings.EqualFold(resultStr, "True") {
		return true, nil
	} else if strings.EqualFold(resultStr, "False") {
		return false, nil
	}
	errl := errors.GenericError
	errl.Message = fmt.Sprintf("SafeEval: Expression %v is not valid. %v", exprObj, resultStr)
	return false, &errl
}
